_posts/2023-01-03-xtdb-phoenix.md

```diff
@@ -0,0 +1,304 @@
+
+Launch an interactive shell and the application by using the command:
+```shell
+iex --erl "-sname xthello@localhost" -S mix phx.server
+```
+
+You should see startup messages and a URL that points you to `http://localhost:4000`.
+You should also see an alert stating that XTDB is not available. That is fine for now
+as we haven't started that service yet. When we do, the application will reconnect to it.
+
+Verify that you have the app up and running by visiting the local URL above.
+Check the Phoenix Framework [Up and Running](https://hexdocs.pm/phoenix/up_and_running.html)
+documentation page for more details.
+
+### Modeling and connecting the database
+
+Next, we want to create the data model we will be storing and querying. We can use the
+person structure we used as a sample earlier, just place that in `lib/model.ex`.
+
+You can type `c "lib/model.ex"` in the interactive shell and the code will be compiled
+and loaded. You can then call `Person.mapping()` to register the mappings for now.
+
+Now that we have everything ready on the Elixir side, we need to go boot up our XTDB node.
+The easiest way to do that is to download a release and run it in demo mode:
+
+```shell
+java -jar xtdberl.jar demo
+```
+
+The demo parameter starts up the node with an in-memory database with default
+mailbox settings that doesn't persist anything. This is good for a quick trial, but for
+actual use, see documentation on how to configure it properly.
+
+After starting up the database, you should see the line ```[notice] 1 XTDB node is now available.```
+in your Phoenix console. We are ready for action!
+
+We can now try inserting some persons in the iex console:
+
+```elixir
+iex> :xt.put(%Person{id: "demo1", first_name: "Demo", last_name: "User", email: "demo@example.com"})
+{:ok, {9, {:timestamp, 1672669602522}}}
+```
+
+### Making a LiveView component
+
+Now that we have modeled the data we want to store and have the database connected, we are
+ready to make a LiveView component that uses it. Let's make a simple person list that has
+a text input that is used to search people by the email field.
+
+First, add the line `Person.mapping()` to `lib/xthello.ex` module to automatically register
+the mappings when we start. Then add the line `live "/people", PeopleLive` inside the `"/"` scope
+in `lib/xthello_web/router.ex`.
+
+Then we create a new file `lib/xthello_web/live/people_live.ex` that implements the basic
+LiveView functions.
+
+```elixir
+defmodule XthelloWeb.PeopleLive do
+  use XthelloWeb, :live_view
+
+  def mount(_params, _session, socket) do
+    {:ok, assign(socket, %{results: nil})}
+  end
+
+  def render(assigns) do
+    ~H"""
+    <div>
+      <.search />
+      <.results results={@results} />
+    </div>
+    """
+  end
+
+  def search(assigns) do
+    ~H"""
+    <input type="text" phx-keyup="update_search" phx-debounce="300" placeholder="Search by email"/>
+    """
+  end
+
+  def results(assigns) do
+    case assigns.results do
```

Comment:
Perhaps comment the pattern matching to happen for the people who dont read easily Elixir code.